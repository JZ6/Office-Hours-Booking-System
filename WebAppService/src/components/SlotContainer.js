import React from "react";
import moment from "moment";

export default class SlotContainer extends React.Component {
	constructor(props) {
		super(props);
		
		this.state = {
			slots: this.props.api.getSlots(this.props.blockId)
		}
		this.prevSlots = this.copySlots(this.state.slots);
		
		this.handleUpdate = this.handleUpdate.bind(this);
		this.handleUndo = this.handleUndo.bind(this);
	}
	
	handleSlotClick = (i) => () => {
		if (this.state.slots[i].utorId === "") {
			// Click on empty slot to assign
			this.editSlot(i, this.props.api.user.utorId, "");
		} else {
			if (this.props.api.user.utorId === this.state.slots[i].utorId) {
				// Click on own slot to delete
				this.editSlot(i, "", "");
			}
			//Click on occupied slot to do nothing
		}
	}
	
	handleUtorIdChange = (i) => (event) => {
		// Delete note to protect privacy
		this.editSlot(i, event.target.value, "");
	}
	
	handleNoteChange = (i) => (event) => {
		// Can't edit note that belongs to nobody
		if (this.state.slots[i].utorId !== "") {
			this.editSlot(i, this.state.slots[i].utorId, event.target.value);
		}
	}
	
	editSlot(i, utorId, note) {
		const newSlots = this.copySlots(this.state.slots);
		newSlots[i].utorId = utorId;
		newSlots[i].note = note;
		this.setState({slots: newSlots});
	}
	
	copySlots(slots) {
		return slots.map(slot => {
			return {utorId: slot.utorId, note: slot.note}
		});
	}
	
	handleSlotConfirm = (i) => () => {
		let updatedSlot = this.props.api.getSlot(this.props.blockId, i);
		if (!this.slotsEqual(this.prevSlots[i], updatedSlot)) {
			// Slot has changed since last update from server
			if (this.props.api.user.role === "student") {
				// TODO: Notify student that this slot has been taken already and abort
				this.editSlot(i, updatedSlot.utorId, updatedSlot.note);
			} else {
				// TODO: Notify instructor that someone has taken this slot and 
				// ask if they want to overwrite that anyway
			}
		} else {
			if (!this.slotsEqual(this.state.slots[i], updatedSlot)) {
				// Slot is different from one on server, post new slot
				this.props.api.postSlot(this.props.blockId, i, this.state.slots[i]);
			}
		}
	}
	
	slotsEqual(a, b) {
		return (a.utorId === b.utorId && a.note === b.note);
	}
	
	handleSlotCancel = (i) => () => {
		this.editSlot(i, this.prevSlots[i].utorId, this.prevSlots[i].note);
	}
	
	handleEmpty = () => {
		// Ensure slots match server's
		this.handleUpdate();
		if (this.props.api.user.role === "student") {
			// Student clears only their own
			for (let i = 0; i < this.state.slots.length; i++) {
				if (this.props.api.user.utorId === this.state.slots[i].utorId) {
					this.editSlot(i, "", "");
					this.props.api.postSlot(this.props.blockId, i, this.state.slots[i]);
				}
			}
		} else {
			// Instructor clears all
			let newSlots = this.state.slots.map((slot, i) => {
				return {utorId: "", note: ""}
			});
			this.setState({slots: newSlots});
			this.props.api.postSlots(this.props.blockId, this.state.slots);
		}
	}
	
	handleUpdate() {
		this.prevSlots = this.props.api.getSlots(this.props.blockId);
		this.setState({slots: this.copySlots(this.prevSlots)});
	}
	
	handleUndo() {
		this.setState({slots: this.copySlots(this.prevSlots)});
	}

	validate(input) {
		return true;
	}
	
	renderUtorId(i) {
		if (this.props.api.user.role === "student") {
			let name;
			if (this.props.api.user.utorId === this.state.slots[i].utorId) {
				// Student owns slot
				name = this.props.api.user.utorId;
			} else {
				if (this.state.slots[i].utorId === "") {
					// Slot is available
					name = "Available";
				} else {
					// Slot is taken by someone else
					name = "Not Available";
				}
			}
			return <div id={`utorId${i}`}>{name}</div>;
		} else {
			return (
				<input
					className="text-input"
					name={`utorId${i}`}
					id={`utorId${i}`}
					type="text"
					value={this.state.slots[i].utorId}
					placeholder="Unassigned UtorID..."
					maxLength={50}
					onChange={this.handleUtorIdChange(i)}
				/>
			);
		}
	}
	
	renderNote(i) {
		if (this.props.api.user.role !== "student" || 
				this.props.api.user.utorId === this.state.slots[i].utorId) {
			// Note editable by instructor or student author
			return (
				<input
					className="text-input"
					name={`note${i}`}
					id={`note${i}`}
					type="text"
					value={this.state.slots[i].note}
					placeholder="Empty Note..."
					maxLength={280}
					onChange={this.handleNoteChange(i)}
				/>
			);
		} else {
			// Note by someone else hidden
			return <div id={`note${i}`} />;
		}
	}
	
	renderSlots() {
		return (
			this.state.slots.map((slot, i) => 
				<div 
					className="slot"
					id={`slot${i}`}
					key={i}
					// Only students can click to assign a free slot to themselves
					onClick={this.props.api.user.role === "student" ? this.handleSlotClick(i) : () => {return false}}
				>
					<button id={`confirm${i}`} onClick={this.handleSlotConfirm(i)}>✎</button>
					<button id={`cancel${i}`} onClick={this.handleSlotCancel(i)}>❌</button>
					{moment(this.props.startTime + this.props.slotDuration * i).format("h:mmA - ")}
					{moment(this.props.startTime + this.props.slotDuration * (i + 1)).format("h:mmA")}
					{this.renderUtorId(i)}
					{this.renderNote(i)}
				</div>
			)
		);
	}

	render() {
		return (
			<div className="slot-container">
				{this.renderSlots()}
				<button id="empty-button" onClick={this.handleEmpty}>
					{this.props.api.user.role === "student" ? (
						"Empty My Slots"
					):(
						"Empty All Slots"
					)}
				</button>
				<button id="update-button" onClick={this.handleUpdate}>Refresh</button>
				<button id="undo-button" onClick={this.handleUndo}>Undo Changes</button>
			</div>
		);
	}
}