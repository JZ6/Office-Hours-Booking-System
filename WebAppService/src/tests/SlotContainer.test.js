import React from "react";
import renderer from 'react-test-renderer';
import {shallow, mount} from "enzyme";
import moment from "moment";
import SlotContainer from "../components/SlotContainer";

function testGetSlots() {
	return [
		{utorId: "parkerpeter15", note: "Everyone gets one."},
		{utorId: "watsonmj25", note: "Face it Tiger... you just hit the jackpot!"},
		{utorId: "", note: ""}
	];
}

let testProps, testSlots, mockPost, mockApi;
beforeEach(() => {
	mockPost = jest.fn();
	mockApi = {
		getSlots: (id) => {return testGetSlots()},
		getSlot: (id, i) => {return testGetSlots()[i]},
		postSlot: mockPost
		};
	testProps = {
		api: mockApi,
		blockId: "someId123",
		startTime: moment("2077/01/01 00:00", "YYYY-MM-DD HH:mm"),
		slotDuration: 600000,
	};
});

describe("instructor view", () => {
	beforeEach(() => {
		mockApi.user = {utorId: "octaviusotto3", role: "instructor"};
	});
	
	test("render", () => {
		const tree = renderer.create(<SlotContainer {...testProps}/>).toJSON();
		expect(tree).toMatchSnapshot();
	});
	
	test("handleSlotClick isn't called", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		wrapper.find("#slot2").props().onClick();
		expect(wrapper.find("#utorId2").text()).toEqual("");
		expect(wrapper.find("#note2").props().value).toEqual("");
	});
	
	test("handleUtorIdChange changes utorId", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		// Edit empty
		wrapper.find("#utorId2").simulate("change", {target: {value: "osbornharry31"}});
		expect(wrapper.find("#utorId2").props().value).toEqual("osbornharry31");
		// Edit occupied (should clear note)
		wrapper.find("#utorId0").simulate("change", {target: {value: "osbornharry31"}});
		expect(wrapper.find("#utorId0").props().value).toEqual("osbornharry31");
		expect(wrapper.find("#note0").props().value).toEqual("");
	});
	
	test("handleNoteChange changes note", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		// Edit empty (shouldn't allow changes)
		wrapper.find("#note2").simulate("change", {target: {value: "Honor is for fools."}});
		expect(wrapper.find("#note2").props().value).toEqual("");
		// Edit occupied
		wrapper.find("#note0").simulate("change", {target: {value: "Honor is for fools."}});
		expect(wrapper.find("#note0").props().value).toEqual("Honor is for fools.");
	});
	
	test("handleSlotConfirm applies slot changes", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		wrapper.find("#utorId0").simulate("change", {target: {value: "osbornharry31"}});
		wrapper.find("#note0").simulate("change", {target: {value: "Honor is for fools."}});
		wrapper.find("#confirm0").simulate("click");
		expect(mockPost).toHaveBeenCalledTimes(1);
		expect(mockPost).toHaveBeenCalledWith("someId123", 0, {utorId: "osbornharry31", note: "Honor is for fools."});
	});
	
	test("handleSlotConfirm handles if slot changed by someone else", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		testProps.api.getSlots = (id) => {
			return [
				{utorId: "parkerpeter15", note: "Everyone gets one."},
				{utorId: "watsonmj25", note: "Face it Tiger... you just hit the jackpot!"},
				{utorId: "parkerpeter15", note: "Sneaky spidey sneaks in."}
			];
		}
		testProps.api.getSlot = (id, i) => {
			return [
				{utorId: "parkerpeter15", note: "Everyone gets one."},
				{utorId: "watsonmj25", note: "Face it Tiger... you just hit the jackpot!"},
				{utorId: "parkerpeter15", note: "Sneaky spidey sneaks in."}
			][i];
		}
		wrapper.find("#utorId2").simulate("change", {target: {value: "osbornharry31"}});
		wrapper.find("#note2").simulate("change", {target: {value: "Honor is for fools."}});
		wrapper.find("#confirm2").simulate("click");
		expect(mockPost).toHaveBeenCalledTimes(0);
		// TODO: Test prompt asking instructor what to do
	});
	
	test("handleSlotCancel cancels changes", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		wrapper.find("#utorId0").simulate("change", {target: {value: "osbornharry31"}});
		wrapper.find("#note0").simulate("change", {target: {value: "Honor is for fools."}});
		wrapper.find("#cancel0").simulate("click");
		expect(wrapper.find("#utorId0").props().value).toEqual("parkerpeter15");
		expect(wrapper.find("#note0").props().value).toEqual("Everyone gets one.");
	});
	
	test("handleSlotConfirm doesn't change if identical to server", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		wrapper.find("#confirm0").simulate("click");
		expect(mockPost).toHaveBeenCalledTimes(0);
	});
	
	test("handleEmpty empties slots", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		wrapper.find("#empty-button").simulate("click");
		expect(wrapper.find("#utorId0").props().value).toEqual("");
		expect(wrapper.find("#note0").props().value).toEqual("");
		expect(wrapper.find("#utorId1").props().value).toEqual("");
		expect(wrapper.find("#note1").props().value).toEqual("");
		expect(wrapper.find("#utorId2").props().value).toEqual("");
		expect(wrapper.find("#note2").props().value).toEqual("");
	});
});

describe("student view", () => {
	beforeEach(() => {
		mockApi.user = {utorId: "parkerpeter15", role: "student"};
	});
	
	test("render", () => {
		const tree = renderer.create(<SlotContainer {...testProps}/>).toJSON();
		expect(tree).toMatchSnapshot();
	});
	
	test("handleSlotClick assigns slot", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		// Click on own slot (unregisters slot and deletes note)
		wrapper.find("#slot0").props().onClick();
		expect(wrapper.find("#utorId0").text()).toEqual("Available");
		expect(wrapper.find("#note0").text()).toEqual("");
		// Click on occupied slot (does nothing)
		wrapper.find("#slot1").props().onClick();
		expect(wrapper.find("#utorId1").text()).toEqual("Not Available");
		expect(wrapper.find("#note1").text()).toEqual("");
		// Click on empty slot (registers slot)
		wrapper.find("#slot2").props().onClick();
		expect(wrapper.find("#utorId2").text()).toEqual("parkerpeter15");
		expect(wrapper.find("#note2").props().value).toEqual("");
	});
	
	test("handleUtorIdChange can't be called", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		expect(wrapper.find("#utorId0").props().onClick).toEqual(undefined);
	});
	
	test("handleNoteChange changes only own notes", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		// Edit empty (shouldn't allow changes)
		expect(wrapper.find("#note2").props().onChange).toEqual(undefined);
		// Edit occupied (shouldn't allow changes)
		expect(wrapper.find("#note1").props().onChange).toEqual(undefined);
		// Edit own
		wrapper.find("#note0").simulate("change", {target: {value: "With great powers comes great responsibility."}});
		expect(wrapper.find("#note0").props().value).toEqual("With great powers comes great responsibility.");
	});
	
	test("handleSlotConfirm applies slot changes", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		wrapper.find("#slot2").simulate("click");
		wrapper.find("#note2").simulate("change", {target: {value: "With great powers..."}});
		wrapper.find("#confirm2").simulate("click");
		expect(mockPost).toHaveBeenCalledTimes(1);
		expect(mockPost).toHaveBeenCalledWith("someId123", 2, {utorId: "parkerpeter15", note: "With great powers..."});
	});
	
	test("handleSlotConfirm handles if slot changed by someone else", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		testProps.api.getSlots = (id) => {
			return [
				{utorId: "parkerpeter15", note: "Everyone gets one."},
				{utorId: "watsonmj25", note: "Face it Tiger... you just hit the jackpot!"},
				{utorId: "osbornharry31", note: "It was me spidey! It was me all along spidey!"}
			];
		}
		testProps.api.getSlot = (id, i) => {
			return [
				{utorId: "parkerpeter15", note: "Everyone gets one."},
				{utorId: "watsonmj25", note: "Face it Tiger... you just hit the jackpot!"},
				{utorId: "osbornharry31", note: "It was me spidey! It was me all along spidey!"}
			][i];
		}
		wrapper.find("#slot2").simulate("click");
		wrapper.find("#note2").simulate("change", {target: {value: "With great powers..."}});
		wrapper.find("#confirm2").simulate("click");
		expect(mockPost).toHaveBeenCalledTimes(0);
		// It updates the slot if student after doing nothing
		expect(wrapper.find("#utorId2").text()).toEqual("Not Available");
		expect(wrapper.find("#note2").text()).toEqual("");
	});
	
	test("handleSlotCancel cancels changes", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		// Edit utorId and note
		wrapper.find("#slot2").simulate("click");
		wrapper.find("#note2").simulate("change", {target: {value: "With great powers..."}});
		wrapper.find("#cancel2").simulate("click");
		expect(wrapper.find("#utorId2").text()).toEqual("Available");
		expect(wrapper.find("#note2").text()).toEqual("");
		// Edit note only
		wrapper.find("#note0").simulate("change", {target: {value: "With great powers..."}});
		wrapper.find("#cancel0").simulate("click");
		expect(wrapper.find("#note0").props().value).toEqual("Everyone gets one.");
	});
	
	test("handleSlotConfirm doesn't change if identical to server", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		wrapper.find("#confirm0").simulate("click");
		expect(mockPost).toHaveBeenCalledTimes(0);
	});
	
	test("handleEmpty empties only own slots", () => {
		const wrapper = shallow(<SlotContainer {...testProps}/>);
		wrapper.find("#empty-button").simulate("click");
		expect(wrapper.find("#utorId0").text()).toEqual("Available");
		expect(wrapper.find("#note0").text()).toEqual("");
		expect(wrapper.find("#utorId1").text()).toEqual("Not Available");
		expect(wrapper.find("#note1").text()).toEqual("");
		expect(wrapper.find("#utorId2").text()).toEqual("Available");
		expect(wrapper.find("#note2").text()).toEqual("");
	});
});
