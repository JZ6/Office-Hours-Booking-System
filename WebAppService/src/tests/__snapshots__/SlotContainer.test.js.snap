// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`instructor view render 1`] = `
<div
  className="slot-container"
>
  <div
    className="slot"
    id="slot0"
    onClick={[Function]}
  >
    <button
      id="confirm0"
      onClick={[Function]}
    >
      ✎
    </button>
    <button
      id="cancel0"
      onClick={[Function]}
    >
      ❌
    </button>
    12:00AM - 
    12:10AM
    <input
      className="text-input"
      id="utorId0"
      maxLength={50}
      name="utorId0"
      onChange={[Function]}
      placeholder="Unassigned UtorID..."
      type="text"
      value="parkerpeter15"
    />
    <input
      className="text-input"
      id="note0"
      maxLength={280}
      name="note0"
      onChange={[Function]}
      placeholder="Empty Note..."
      type="text"
      value="Everyone gets one."
    />
  </div>
  <div
    className="slot"
    id="slot1"
    onClick={[Function]}
  >
    <button
      id="confirm1"
      onClick={[Function]}
    >
      ✎
    </button>
    <button
      id="cancel1"
      onClick={[Function]}
    >
      ❌
    </button>
    12:10AM - 
    12:20AM
    <input
      className="text-input"
      id="utorId1"
      maxLength={50}
      name="utorId1"
      onChange={[Function]}
      placeholder="Unassigned UtorID..."
      type="text"
      value="watsonmj25"
    />
    <input
      className="text-input"
      id="note1"
      maxLength={280}
      name="note1"
      onChange={[Function]}
      placeholder="Empty Note..."
      type="text"
      value="Face it Tiger... you just hit the jackpot!"
    />
  </div>
  <div
    className="slot"
    id="slot2"
    onClick={[Function]}
  >
    <button
      id="confirm2"
      onClick={[Function]}
    >
      ✎
    </button>
    <button
      id="cancel2"
      onClick={[Function]}
    >
      ❌
    </button>
    12:20AM - 
    12:30AM
    <input
      className="text-input"
      id="utorId2"
      maxLength={50}
      name="utorId2"
      onChange={[Function]}
      placeholder="Unassigned UtorID..."
      type="text"
      value=""
    />
    <input
      className="text-input"
      id="note2"
      maxLength={280}
      name="note2"
      onChange={[Function]}
      placeholder="Empty Note..."
      type="text"
      value=""
    />
  </div>
  <button
    id="empty-button"
    onClick={[Function]}
  >
    Empty All Slots
  </button>
  <button
    id="update-button"
    onClick={[Function]}
  >
    Refresh
  </button>
  <button
    id="undo-button"
    onClick={[Function]}
  >
    Undo Changes
  </button>
</div>
`;

exports[`student view render 1`] = `
<div
  className="slot-container"
>
  <div
    className="slot"
    id="slot0"
    onClick={[Function]}
  >
    <button
      id="confirm0"
      onClick={[Function]}
    >
      ✎
    </button>
    <button
      id="cancel0"
      onClick={[Function]}
    >
      ❌
    </button>
    12:00AM - 
    12:10AM
    <div
      id="utorId0"
    >
      parkerpeter15
    </div>
    <input
      className="text-input"
      id="note0"
      maxLength={280}
      name="note0"
      onChange={[Function]}
      placeholder="Empty Note..."
      type="text"
      value="Everyone gets one."
    />
  </div>
  <div
    className="slot"
    id="slot1"
    onClick={[Function]}
  >
    <button
      id="confirm1"
      onClick={[Function]}
    >
      ✎
    </button>
    <button
      id="cancel1"
      onClick={[Function]}
    >
      ❌
    </button>
    12:10AM - 
    12:20AM
    <div
      id="utorId1"
    >
      Not Available
    </div>
    <div
      id="note1"
    />
  </div>
  <div
    className="slot"
    id="slot2"
    onClick={[Function]}
  >
    <button
      id="confirm2"
      onClick={[Function]}
    >
      ✎
    </button>
    <button
      id="cancel2"
      onClick={[Function]}
    >
      ❌
    </button>
    12:20AM - 
    12:30AM
    <div
      id="utorId2"
    >
      Available
    </div>
    <div
      id="note2"
    />
  </div>
  <button
    id="empty-button"
    onClick={[Function]}
  >
    Empty My Slots
  </button>
  <button
    id="update-button"
    onClick={[Function]}
  >
    Refresh
  </button>
  <button
    id="undo-button"
    onClick={[Function]}
  >
    Undo Changes
  </button>
</div>
`;
